File: Qual/1_21_6/13_21_1_QE_prob3.py
class GNode:
    def __init__(self, id, color="W", d=-1, f=-1, p=None):
        self.id = id # id is a string
        self.color = color # color (status) of node
        self.d = d # discover time of node
        self.f = f # finish time of node
        self.parent = p # predecessor time of node

    def __str__(self):
        return self.id  

def merge(L:list, first:int, mid:int, last:int) -> None:
    k = first
    sub1 = L[first:mid+1]
    sub2 = L[mid+1:last+1]
    i = j = 0
    while i < len(sub1) and j < len(sub2):
        if sub1[i].id <= sub2[j].id:
            L[k] = sub1[i]
            i = i + 1
        else:
            L[k] = sub2[j]
            j = j + 1
        k = k + 1
    if i < len(sub1):
        L[k:last+1] = sub1[i:]
    elif j < len(sub2):
        L[k:last+1] = sub2[j:]


def mergeSortHelp(L:list, first: int, last: int)->None:
    if first == last:
        return
    else:
        mid = first + (last - first) // 2
        mergeSortHelp(L, first, mid)
        mergeSortHelp(L, mid+1, last)
        merge(L, first, mid, last)

def mergeSort(L:list)->None:
    mergeSortHelp(L, 0, len(L)-1)



def LexSmallest(G: dict, t:GNode, v:GNode)->list:
    out = []
    if t == v: return [t.id]
    find = 0
    def dfs(st:GNode, en:GNode)->None:
        nonlocal find # 이걸 붙여줘야 바깥 scope와 연계 가능.
        if find: return
        st.color = "K" # "K" means visited, "W" means not visitied;
        out.append(st.id)
        for nxt in G[st]:
            if(nxt.color == "K"): continue
            nxt.color = "K"
            if(nxt == en):
                find = 1
                out.append(nxt.id)
                return
            dfs(nxt, en)
    for n in G:
        # mergeSort(G[n])
        G[n].sort(key = lambda x: x.id)
    dfs(t, v)
    if find:
        return out
    else:
        return []


if __name__ == "__main__":
    r, s, t, u = GNode('r'), GNode('s'), GNode('t'), GNode('u')
    v, w, x, y = GNode('v'), GNode('w'), GNode('x'), GNode('y')
    G = dict()
    G[r], G[w], G[t], G[u] = [w, v], [s, r, t], [w, s, x], [y]
    G[v], G[s], G[x], G[y] = [r], [w, t, x], [s, t], [u]

    print(LexSmallest(G, t, v))
    print(LexSmallest(G, u, u))
    print(LexSmallest(G, x, y))

File: Qual/1_21_6/13_21_1_QE_prob3.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#define MAX_NODES 100

typedef struct GNode {
    char id;
    char color;
    int d;
    int f;
    struct GNode* parent;
} GNode;

GNode* createGNode(char id) {
    GNode* newNode = (GNode*) malloc(sizeof(GNode));
    newNode->id = id;
    newNode->color = 'W';
    newNode->d = -1;
    newNode->f = -1;
    newNode->parent = NULL;
    return newNode;
}

bool find = false;
char output[MAX_NODES];
int outputIndex = 0;

void dfs(GNode* G[MAX_NODES][MAX_NODES], GNode* st, GNode* en) {
    if (find) return;

    st->color = 'K';
    output[outputIndex++] = st->id;

    for (int i = 0; i < MAX_NODES; i++) {
        if (G[(int)(st->id - 'a')][i] == NULL || G[(int)(st->id - 'a')][i]->color == 'K') continue;

        G[(int)(st->id - 'a')][i]->color = 'K';
        if (G[(int)(st->id - 'a')][i] == en) {
            find = true;
            output[outputIndex++] = en->id;
            return;
        }
        dfs(G, G[(int)(st->id - 'a')][i], en);
    }
}

char* lexSmallest(GNode* G[MAX_NODES][MAX_NODES], GNode* t, GNode* v) {
    find = false;
    memset(output, 0, sizeof(output));
    outputIndex = 0;
    dfs(G, t, v);
    return output;
}

int main() {
    GNode* r = createGNode('r');
    GNode* s = createGNode('s');
    GNode* t = createGNode('t');
    GNode* u = createGNode('u');
    GNode* v = createGNode('v');
    GNode* w = createGNode('w');
    GNode* x = createGNode('x');
    GNode* y = createGNode('y');

    GNode* G[MAX_NODES][MAX_NODES] = {0};

    G[(int)(r->id - 'a')][0] = w;
    G[(int)(r->id - 'a')][1] = v;
    G[(int)(w->id - 'a')][0] = s;
    G[(int)(w->id - 'a')][1] = r;
    G[(int)(w->id - 'a')][2] = t;
    G[(int)(t->id - 'a')][0] = w;
    G[(int)(t->id - 'a')][1] = s;
    G[(int)(t->id - 'a')][2] = x;
    G[(int)(u->id - 'a')][0] = y;
    G[(int)(v->id - 'a')][0] = r;
    G[(int)(s->id - 'a')][0] = w;
    G[(int)(s->id - 'a')][1] = t;
    G[(int)(s->id - 'a')][2] = x;
    G[(int)(x->id - 'a')][0] = s;
    G[(int)(x->id - 'a')][1] = t;
    G[(int)(y->id - 'a')][0] = u;

    printf("%s\n", lexSmallest(G, t, v));
    printf("%s\n", lexSmallest(G, u, u));
    printf("%s\n", lexSmallest(G, x, y));

    free(r);
    free(s);
    free(t);
    free(u);
    free(v);
    free(w);
    free(x);
    free(y);

    return 0;
}
